src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -1904,6 +2189,154 @@ void sort(json::Value& val) {
+    // receive some valid but nonsensical edge cases. these are not supported
+    // but do not generate errors here, since they are problematic only when a
+    // accessed by a $ref. so we don't register them in the base_uris_map, an
+    // will raise an error later if we try to access them.
+    // keyword | $schema  | is bundled schema
+    //   "$id" |          | if root is >=draft6
+    //   "$id" | >=draft6 |       yes
+    //   "$id" |  draft4  |       no
+    //   "id"  |          | if root is draft4
+    //   "id"  | >=draft6 |       no
+    //   "id"  |  draft4  |       yes
+
+    auto maybe_draft4_id_it = this_obj.find("id");
+    auto maybe_id_it = this_obj.find("$id");
+    if (
+      maybe_id_it != this_obj.object_range().end()
+      || maybe_draft4_id_it != this_obj.object_range().end()) {
+        // we are visiting a bundled schema. the dialect has to be known and has
+        // to match the keyword used.
+        // try to extract the dialect from the $schema keyword, or use the
+        // parent dialect. empty means that "$schema" is present but the dialect
+        // is not known, and we should stop scanning this branch.
+        auto maybe_new_dialect = [&]() -> std::optional<json_schema_dialect> {
+            auto dialect_it = this_obj.find("$schema");
+            if (dialect_it == this_obj.object_range().end()) {
+                // we can keep using the parent dialect
+                return dialect;
+            }
+
+            // we have a $schema keyword, use this dialect if we find out that
+            // this_obj is a bundled schema
+            return from_uri(dialect_it->value().as_string_view());
+        }();
+
+        if (maybe_new_dialect.has_value() == false) {
+            // stop scanning this tree, we might be in a bundled schema but we
+            // don't know the dialect.
+            return;
+        }
+
+        // we are in a bundled schema and we know the dialect to use, now we
+        // know which keyword to use to get the base_uri
+        auto id_it = [&] {
+            switch (maybe_new_dialect.value()) {
+            case json_schema_dialect::draft4:
+                return maybe_draft4_id_it;
+            case json_schema_dialect::draft6:
+            case json_schema_dialect::draft7:
+            case json_schema_dialect::draft201909:
+            case json_schema_dialect::draft202012:
+                return maybe_id_it;
+            }
+        }();
+
+        if (id_it == this_obj.object_range().end()) {
+            // stop scanning this branch, the keyword for base uri does not
+            // agree with schema dialect.
+            return;
+        }
+
+        // run validation since we are not a guaranteed to be in proper schema
+        if (validate_json_schema(maybe_new_dialect.value(), this_obj)
+              .has_error()) {
+            // stop exploring this branch, the schema is invalid
+            return;
+        }
```

Comment:
checking the ref implementation behavior, will add a commit to match the result

Reply:
taking the stricter approach to throw if the bundled schema is invalid.

in a follow up this can be relaxed if the bundled is never actually referenced